File: Forge/src/main/java/ram/talia/moreiotas/forge/eventhandlers/ChatEventHandler.java
package ram.talia.moreiotas.forge.eventhandlers;

import net.minecraft.world.entity.player.Player;
import net.minecraftforge.event.ServerChatEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import org.jetbrains.annotations.Nullable;
import ram.talia.moreiotas.api.mod.MoreIotasConfig;
import ram.talia.moreiotas.api.util.ChatEntry;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class ChatEventHandler {
    private static final String TAG_CHAT_PREFIX = "moreiotas:prefix";

    private static final Map<UUID, @Nullable String> lastMessages = new HashMap<>();
    private static final Map<UUID, Long> lastMessageTimestamps = new HashMap<>();

    private static final ArrayDeque<ChatEntry> messageLog = new ArrayDeque<>();

    private static long lastMessageTimestamp = 0;
    private static int messagesHandled = 0;

    public static void setPrefix(Player player, @Nullable String prefix) {
        if (prefix == null)
            player.getPersistentData().remove(TAG_CHAT_PREFIX);
        else
            player.getPersistentData().putString(TAG_CHAT_PREFIX, prefix);
    }

    public static @Nullable String getPrefix(Player player) {
        if (!player.getPersistentData().contains(TAG_CHAT_PREFIX))
            return null;
        return player.getPersistentData().getString(TAG_CHAT_PREFIX);
    }

    public static @Nullable String getLastMessage(@Nullable Player player) {
        if (player != null)
            return lastMessages.get(player.getUUID());
        if (!messageLog.isEmpty())
            return messageLog.peekLast().message();
        return null;
    }


	public static long lastMessageTimestamp(@Nullable Player player) {
        if (player != null)
            return lastMessageTimestamps.getOrDefault(player.getUUID(), 0l);
        return lastMessageTimestamp;
	}

	public static int lastMessageCount() {
        return messagesHandled;
	}

	public static List<ChatEntry> chatLog(int count) {
        var list = new ArrayList<ChatEntry>();
        var iter = messageLog.descendingIterator();
        for (int i = 1; i < count; i++) {
            if (!iter.hasNext())
                break;
            list.add(iter.next());
        }
        return list;
    }

    @SubscribeEvent
    public static void chatMessageSent(ServerChatEvent event) {
        if (event.isCanceled())
            return;

        var player = event.getPlayer();
        var uuid = player.getUUID();
        var timestamp = player.serverLevel().getGameTime(); // Dunno why it wants closeable try
        var text = event.getRawText();

        var prefix = ChatEventHandler.getPrefix(player);

        if (prefix != null && text.startsWith(prefix)) {
            event.setCanceled(true);
            lastMessages.put(uuid, text.substring(prefix.length()));
            lastMessageTimestamps.put(uuid, timestamp);
            return;
        }

        if (prefix == null) {
            lastMessages.put(uuid, text);
            lastMessageTimestamps.put(uuid, timestamp);
        }

        if (MoreIotasConfig.getServer().getMaxChatLog() == 0)
            return;

        while (!messageLog.isEmpty() && messageLog.size() > MoreIotasConfig.getServer().getMaxChatLog()) {
            messageLog.removeFirst();
        }

        messageLog.addLast(new ChatEntry(text, timestamp, event.getPlayer().getName().getString()));

        if (lastMessageTimestamp == timestamp) {
            messagesHandled++;
        } else {
            messagesHandled = 1;
            lastMessageTimestamp = timestamp;
        }
    }
}

File: Fabric/src/main/java/ram/talia/moreiotas/fabric/eventhandlers/ChatEventHandler.kt
package ram.talia.moreiotas.fabric.eventhandlers

import net.minecraft.network.chat.ChatType
import net.minecraft.network.chat.PlayerChatMessage
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player
import ram.talia.moreiotas.api.util.ChatEntry
import ram.talia.moreiotas.fabric.cc.MoreIotasCardinalComponents
import java.util.UUID
import kotlin.collections.ArrayDeque

object ChatEventHandler {
    private val lastMessages: MutableMap<UUID, String?> = mutableMapOf()
    private val messageLog: ArrayDeque<ChatEntry> = ArrayDeque<ChatEntry>()
    private var lastMessageTimestamp: Long? = null;
    private var messagesHandled: Int = 0;

    private val lastMessageTimestamps: MutableMap<UUID, Long?> = mutableMapOf()

    fun receiveChat(message: PlayerChatMessage, player: ServerPlayer, params: ChatType.Bound): Boolean {
        val text = message.signedBody.content + (message.unsignedContent?.string ?: "")

        val prefix = MoreIotasCardinalComponents.CHAT_PREFIX_HOLDER.get(player).prefix

        var timestamp = player.serverLevel().gameTime

        if (prefix != null && text.startsWith(prefix)) {
            lastMessages[player.uuid] = text.substring(prefix.length)
            lastMessageTimestamps[player.uuid] = timestamp
            return false
        }

        if (prefix == null) {
            lastMessages[player.uuid] = text
            lastMessageTimestamps[player.uuid] = timestamp
        }

        while (messageLog.size > 1023) {
            // Just in case somehow we end up putting two "at once" even though this is only place this is touched
            messageLog.removeFirst()
        }

        messageLog.addLast(ChatEntry(text, timestamp, player.getName().getString()))
        if (lastMessageTimestamp == timestamp) {
            messagesHandled++
        } else {
            messagesHandled = 1
            lastMessageTimestamp = timestamp
        }
        return true
    }

    @JvmStatic
    fun lastMessage(player: Player?): String? = if (player != null) lastMessages[player.uuid] else messageLog.last().message

    @JvmStatic
    fun lastMessageTimestamp(player: Player?): Long? = if (player != null) lastMessageTimestamps[player.uuid] else lastMessageTimestamp

    @JvmStatic
    fun chatLog(count: Int): List<ChatEntry> {
        return messageLog.subList(messageLog.size - count, messageLog.size)
    }
}